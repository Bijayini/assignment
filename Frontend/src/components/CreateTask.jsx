import React, {Component} from "react";
import PropTypes from 'prop-types';

import Modal from "../ui-components/Modal";
import Loader from "../ui-components/Loader";

import TaskForm from "../components/TaskForm";
import Error from "../pages/Error";
import labels from '../labels/Task.labels';

class CreateTask extends Component{
    static propTypes = {
        toggleOverlay: PropTypes.func.isRequired,
        loading:PropTypes.bool,
        error:PropTypes.string,
    };

    constructor(props){
        super(props);
        this.state = {
            title: "",
            description: "",
            status:'',
            hasError:false
        };
    }


    handleChange =(event)=> {
        const target = event.target;
        const value = target.value;
        const name = target.name;

        this.setState({
            [name]: value
        });
    };

    handleSubmit = () => {
        const {title, description, status} = this.state;
        const {createTask, toggleOverlay } = this.props;

        if(title !== '' && description !== '' && status !== ''){
            createTask({
                title,
                description,
                status
            });
            toggleOverlay();
        }else{
            this.setState({hasError:true});
        }
    };

    render(){
        const { error, loading,toggleOverlay} = this.props;
        const {title, description, status, hasError} = this.state;
        const buttons = [
            {
                label:labels.cancel,
                type: 'secondary',
                callback:toggleOverlay,
            },
            {
                label: labels.continue,
                type: 'primary',
                callback:this.handleSubmit,
            },
        ];
        if (error) {
            return <Error message={error}/>
        }

        if (loading) {
            return <Loader/>;
        }
        return(
            <Modal toggleOverlay={toggleOverlay} buttons={buttons}>
                <h3 className="modal-title">{labels.create_a_task}</h3>
                <TaskForm title={title} description={description} status={status} handleChange={this.handleChange} hasError={hasError}/>
            </Modal>
        );
    }
}
export default  CreateTask;