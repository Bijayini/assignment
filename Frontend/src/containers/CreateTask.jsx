import React, { Component } from "react";
import { connect } from "react-redux";
import PropTypes from 'prop-types';

import Modal from "../ui-components/Modal";

import { fetchTasks } from "../actions";


class CreateTask extends Component{
  static propTypes = {
    toggleOverlay: PropTypes.func.isRequired,
  };


  constructor(props){
    super(props);
  }

  render(){
    const {toggleOverlay} = this.props;
    const buttons = [
      {
        label: 'cancel',
        type: 'secondary',
        callback:toggleOverlay,
      },
      {
        label: 'continue',
        type: 'primary',
        callback: ()=>{},
      },
    ];
    return(
      <Modal toggleOverlay={toggleOverlay} buttons={buttons}>
        <h2>Create a Task</h2>
        <form>
          <div className="form-field">
            <label htmlFor="title">Title</label>
            <input type="text" id="title"/>
          </div>
          <div className="form-field">
            <label htmlFor="description">Description</label>
            <textarea id="description" />
          </div>
          <div className="form-field">
            <label htmlFor="status">Status</label>
            <select>
              <option value="in progress">In Progress</option>
              <option value="completed">Completed</option>
              <option value="urgent">Urgent</option>
            </select>
          </div>
        </form>
      </Modal>
    );
  }
}

const mapStateToProps = state => ({
  tasks: state.tasks,
});

const mapDispatchToProps = dispatch => {
  return {
    fetchTasks: () => {
      dispatch(fetchTasks());
    },
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(
  CreateTask,
);
