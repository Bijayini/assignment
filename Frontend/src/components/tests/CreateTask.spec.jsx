import React from 'react';
import { shallow } from 'enzyme';
import CreateTask from '../CreateTask';


describe('CreateTask', () => {
  let component;
  let props;

  beforeEach(()=>{
    props = {
      error:null,
      loading:false,
      toggleOverlay:jest.fn(),
      createTask:jest.fn()
    };
    component = shallow(<CreateTask {...props}/>);
  });

  it('should match with snapshot', () => {
    expect(component.debug()).toMatchSnapshot();
  });

  it('should render error component when error is not null', ()=>{
    props.error = 'some error';
    component = shallow(<CreateTask {...props}/>);

    expect(component.find('Loader').length).toBe(0);
    expect(component.find('TaskForm').length).toBe(0);
    expect(component.find('Error').length).toBe(1);
  });

  it('should render loader component when loading is true', ()=>{
    props.loading = true;
    component = shallow(<CreateTask {...props}/>);

    expect(component.find('Loader').length).toBe(1);
    expect(component.find('TaskForm').length).toBe(0);
    expect(component.find('Error').length).toBe(0);
  });

  it('should update title on handleChange', ()=>{
    component.instance().handleChange({target:{name:'title', value:'some value'}});

    expect(component.instance().state.title).toBe('some value');
  });

  it('should not call createTask if any form field is empty on handleSubmit', ()=>{
    component.instance().handleSubmit();

    expect(component.instance().state.hasError).toBe(true);
    expect(props.createTask).not.toHaveBeenCalled();
  });

  it('should call createTask if all form fields are having value', ()=>{
    component.setState({
      title:'some title',
      description:'some desc',
      status:'some value'
    });
    component.instance().handleSubmit();

    expect(component.instance().state.hasError).toBe(false);
    expect(props.createTask).toHaveBeenCalledWith({description: "some desc", status: "some value", title: "some title"});
  })
});
