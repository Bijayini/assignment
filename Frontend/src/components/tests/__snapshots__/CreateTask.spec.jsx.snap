// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateTask should match with snapshot 1`] = `
"<Modal toggleOverlay={[Function: mockConstructor]} buttons={{...}}>
  <h3 className=\\"modal-title\\">
    Create A Task
  </h3>
  <TaskForm title=\\"\\" description=\\"\\" status=\\"\\" handleChange={[Function]} hasError={false} />
</Modal>"
`;
